STYLEGUIDE.md

```diff
@@ -0,0 +1,716 @@
+
+## White space
+
+### Indentation: Use 2 spaces, not tabs ✓
+Use 2 spaces for indentation in all code files, unless not possible for some reason.
+Don't use tabs.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use 4 spaces */
+arr.forEach(item => {
+····const property = item.property;
+});
+
+/* Don't use tabs */
+arr.forEach(item => {
+⇥const property = item.property;
+});
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use 2 spaces */
+arr.forEach(item => {
+··const property = item.property;
+});
+```
+
+### Add one empty line before and after function definitions
+
+Add exactly one empty line before and after every function definition.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't add no lines */
+function foo() {
+  ...
+}
+function bar() {
+  ...
+}
+
+/* Don't add more than one line */
+function foo() {
+  ...
+}
+
+
+function bar() {
+  ...
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Add one line */
+function foo() {
+  ...
+}
+
+function bar() {
+  ...
+}
+```
+
+## Comments
+
+### Document public API with TypeDoc comments
+API that is exposed publicly, should be fully documented.
+TypeDoc comments should start with `/**`.
+Functions that return something should specify the return value with `@returns`.
+Parameters should not have type annotations as those can be inherently retrieved from the TypeScript code.
+Documentation comments should have the following structure:
+1. If necessary, one line for deprecation.
+2. Description, that can extend over multiple lines.
+3. If necessary, one line per type parameter.
+4. One line per parameter.
+5. If necessary, one line for the description of the return value.
+
+There should be no empty lines in between, except if it is part of the description.
+
+### Use `@deprecated` tag for deprecation
+When deprecating public API, this is done through the TypeDoc comments.
+The first line of the comment should start with `@deprecated` followed by a note since which version this is deprecated (e. g. `Since v1.0.0.`) and a note what to use instead (or alternatively that it won't be replaced).
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't leave out from which version on this is deprecated */
+/**
+ * @deprecated Use [[foo]] instead.
+ */
+
+/* Don't leave out what to use instead */
+/**
+ * @deprecated Since v1.2.3.
+ */
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Add all deprecation information */
+/**
+ * @deprecated Since v1.2.3. Use [[foo]] instead.
+ */
+```
+
+### Don't reference types in parameters an type parameters
+In JSDoc it is common to specify the types of parameters and type parameters as JavaScript has not types.
+In TypeDoc this is not necessary, because TypeScript inherently has types and those should not differ between the implementation and documentation.
+A parameter or type parameter line should start with `@param` or `@typeparam`, followed by the (type) parameter name, a dash and the description.
+The description should end with a fullstop.
```

Comment:
```suggestion
### Don't reference types in parameters an type parameters
In JSDoc it is common to specify the types of parameters and type parameters as JavaScript has no types.
In TypeDoc this is not necessary, because TypeScript inherently has types and those should not differ between the implementation and documentation.
A parameter or type parameter line should start with `@param` or `@typeparam`, followed by the (type) parameter name, a dash, and the description.
The description should end with a full stop.
```